Modules/arraymodule.c

```diff
@@ -1115,18 +1115,39 @@ array_array_count(arrayobject *self, PyObject *v)
 array.array.index
 
     v: object
+    start: slice_index(accept={int}) = 0
+    stop: slice_index(accept={int}, c_default="PY_SSIZE_T_MAX") = sys.maxsize
     /
 
 Return index of first occurrence of v in the array.
+
+Raises ValueError if the value is not present.
 [clinic start generated code]*/
 
 static PyObject *
-array_array_index(arrayobject *self, PyObject *v)
-/*[clinic end generated code: output=d48498d325602167 input=cf619898c6649d08]*/
+array_array_index_impl(arrayobject *self, PyObject *v, Py_ssize_t start,
+                       Py_ssize_t stop)
+/*[clinic end generated code: output=c45e777880c99f52 input=6efdb3dae558556c]*/
 {
-    Py_ssize_t i;
+    Py_ssize_t i, len;
 
-    for (i = 0; i < Py_SIZE(self); i++) {
+    len = Py_SIZE(self);
+
+    if (start < 0) {
+        start += len;
+
+        if (start < 0) {
+            start = 0;
+        }
+    }
+
+    if (stop < 0) {
+        stop += len;
+    } else if (stop > len) {
```

Comment:
"else if" should be on a separate line (PEP 7). But actually this branch is not needed due to the comment below.